### DISCUSSION TITLE
Render styled linking annotations

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#13] [How Might We]
(WIP)

In any canvas rendering scenario, if the canvas has linking annotations available, render them as hyperlinks on the image surface.

Example:

```
{
    "id": "http://example.org/anno-l",
    "type": "Annotation",
    "motivation": ["linking"],
    "body": "https://en.wikipedia.org/wiki/The_Medallion",
    "target": {
        "id": "https://data.ng-london.org.uk/iiif/0CWR-0001-0000-0000/canvas/-1#xywh=10000,10000,2000,2000",
        "type": "Canvas",
        "partOf": [
            {
                "id": "https://data.ng-london.org.uk/iiif/0CWR-0001-0000-0000/manifest",
                "type": "Manifest"
            }
        ]
    }
}
```

(NB the `target` of the anno is not the target of the link, the anno target is the relevant part of the canvas, and the body of the anno is used to generate the appropriate href. Here it's just a link but it could be an object:

``` 
{
    "id": "https://en.wikipedia.org/wiki/The_Medallion",
    "type": "Text",
    "format": "text/html"
}
```

Questions

This is an example of where the linking annotation might be in a separate web resource - linked from the Canvas in the manifest via the `annotations` property. The component would need to follow that link to find out what annotations are available.

This opens up a world of complexity.

 - need to be able to specify whether component should follow `annotations` links at all
 - it could find anything on the other ends of those links. See the discussion in #10 (Choice) for discussion of a pattern, which should be common across multiple types of resource, for the developer's code to learn about what annos/choices/etc are available, and then to do something with them. Which might be a reactive, data-binding kind of pattern, but might also be a procedural, interrogative followed by imperative pattern - "what have you got? OK, do _this_ with _that_"
 - (break this pattern discussion out into separate HMW?)
 - Need to style the link shapes
 - Might need to render text
 - There are common patterns here for other types of annotation that associate a region of the page with some web resource. A `tagging` anno looks very like a `linking` anno and you might want identical behaviour in some applications (e.g., tag things with a vocabulary URL that resolves to a topic page).
 - The developer can also connect a canvas with some annos at code-time - annos that aren't linked from the source manifest.
 - (make this into a more general HMW?)

**Comment by stephenwf** [comment]
@tomcrane is there a cookbook for this type of annotation, or could you provide an example manifest (maybe in a gist).

**Reply by tomcrane** [reply]
https://digirati-co-uk.github.io/iiif-canvas-panel/extra-fixtures/linking.json

**Comment by stephenwf** [comment]
On the data side we need something in the tools to detect wether an annotation is a linking, or extract linking annotations from a page.

```js
if (SomeHelper.isLinkingAnnotation(annotation)) {
  // ...
}
```

And then some wrapper or abstraction over the linking annotation which will make working with it easier.

```js
const linking = new LinkingAnnotation(annotation);

linking.getSelector(); // x, y, width, height
linking.getType(); // type..
linking.isManifest(); // is remote?
linking.isLocalCanvas(); //  ??
// ...

```

Or a normalised data structure:
```js
const linking = normaliseLinkingAnnotation(annotation);

linking.selector // x, y, width, height
linking.isManifest // true / false
linking.type // webpage / manifest / etc.

```

**Reply by tomcrane** [reply]
`.selector` in the above is I think the Target helper class.
For a linking anno the body is the destination of the link.
Will explore this in the proposal below.

**Comment by tomcrane** [comment]
Proposal

There are different scenarios here.

One is similar to explicit highlighting, as in #12 - the only difference being that the target on the canvas is a clickable link. 
Here we are adding a linking annotation to the canvas:

```html

<canvas-panel id="cp"></canvas-panel>
<script>
   const cp = document.getElementById("cp");
   const vault = .. // where do we get the shared vault from in this scenario?
   await vault.loadManifest("..manifest_id..");
   cp.setCanvas("..canvas_id..");

  // this will be the clickable shape
  // we don't need to specify the canvas that this shape is on, we're going to add it to cp in a moment.
  // cp <=> canvas so it can only target cp's current canvas.
  const linkBox = new Target("xywh=2000,2000,1000,1000");
 ​
  ​// create an annotation with motivation "linking"
  ​const anno = new LinkingAnnotation();
  ​anno.target = linkBox;

  // we'll come back to what this might be later
  anno.body = "https://artgallery.yale.edu/collections/objects/34001"; 
 
  ​// create anno in vault
  ​vault.load("my-anno", anno);

  // TODO - still think of a better name than DisplayAnnotation - 
  // it's a visual and sometimes interactive _thing_ on the canvas.
  // compare this with #12 - we're constructing our DisplayAnnotation
  ​const displayAnno = new DisplayAnnotation(vault.FromRef("my-anno"));
  ​displayAnno.cssClass = "link-shape";
      ​
  ​cp.displayAnnotations.add(displayAnno);
  
</script> 
```

By default here, because the body of the anno is a bare link, cp can turn this into a hyperlink.

But what if the body is a canvas within a manifest? Even if CP recognises that, it doesn't know what you want to do with it as a link - e.g., turn it into a link to some other viewer with a content state on the query string.

How about you can provide a function to process the body?

```js

// cp will call this. The Body class is like Target - it's not a raw W3C anno body, 
// it's a wrapper with helpers.
function getLinkFromBody(body) {
    const cs = body.toContentState();
    if(cs) {
       return "https://example.org/viewer?iiif-content=" + cs.getEncoded();
    }
    return body.toUrl(); // TBC...
}

// this will be a linking anno...
const options = {
   linkFromBody: getLinkFromBody,
   linkCssClass: "link-shape"
}
​const displayAnno = new DisplayAnnotation(vault.FromRef("my-anno"), options);

```

The second scenario is where the annotations are already present in the IIIF resource, or linked from it via the canvas `annotations` property.

Canvas panel's default behaviour is (via vault) to follow `annotations`. So it has raw W3C annos available. Then we need to handle them. (Does CP have them? What does that mean - that they are loaded into the vault? They aren't drawn yet).

It can do this the same way manually added links, highlights etc are done, with DisplayAnnotation, but that needs some assistance and control.

Do we ever just allow it to attempt to display all the annotations?

The painting annos, inline in the canvas (usually, but might require a dereference of the annotation page), should always be displayed. They are the visible (and/or audible) content of the canvas.

Other annotations are linked via the `annotations` property.

If they have the motivation `supplementing`, they will be fed down a different path. They are possibly transcriptions, OCR text, captions, etc. Handle this somewhere else. See #15 

All other annotations could just be each wrapped with DisplayAnnotation instances and cp would attempt to draw them

```html
<!-- prevent default follow behaviour -->
<canvas-panel id="cp" follow-annotations="false" />


<!-- allow cp default behaviour in anno rendering, but at least provide some styles -->
<canvas-panel id="cp"
    highlight-css-class="anno-hilite"
    link-css-class="anno-link"    
/>
```

WIP
